src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/AutoFlushingStream.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor
+{
+    internal class AutoFlushingStream : Stream
+    {
+        private readonly Stream _inner;
+
+        public AutoFlushingStream(Stream inner)
+        {
+            _inner = inner;
+        }
+
+        public override bool CanRead => _inner.CanRead;
+
+        public override bool CanSeek => _inner.CanSeek;
+
+        public override bool CanWrite => _inner.CanWrite;
+
+        public override long Length => _inner.Length;
+
+        public override long Position { get => _inner.Position; set => _inner.Position = value; }
+
+        public override long Seek(long offset, SeekOrigin origin) => _inner.Seek(offset, origin);
+
+        public override void SetLength(long value) => _inner.SetLength(value);
+
+        public override void Flush() => FlushAsync(CancellationToken.None).GetAwaiter().GetResult();
```

Comment:
You're correct, I just adapted that logic so the synchronization is managed on the `AutoFlushingStream` side of things rather than the `NerdbankStreams`. The original implementation relied on `NerdbankStreams` to handle synchronization. This synchronization happened in multiple parts, including the `PipeStream` as well as the individual stream readers/writers. Something was being compromised in the process resulting in concurrent read/write exceptions.

https://github.com/AArnott/Nerdbank.Streams/blob/master/src/Nerdbank.Streams/PipeStream.cs#L238-L247 uses the same essential logic.